display_list/display_list_path_effect_unittests.cc

```diff
@@ -0,0 +1,223 @@
+  ASSERT_NE(path_effect.shared().get(), &path_effect);
+  ASSERT_EQ(*path_effect.shared(), path_effect);
+}
+
+TEST(DisplayListPathEffect, UnknownContents) {
+  const SkScalar TestDashes1[] = {4.0, 2.0};
+  sk_sp<SkPathEffect> sk_effect = SkDashPathEffect::Make(TestDashes1, 2, 0.0);
+  DlUnknownPathEffect effect(sk_effect);
+  ASSERT_EQ(effect.skia_object(), sk_effect);
+  ASSERT_EQ(effect.skia_object().get(), sk_effect.get());
+}
+
+TEST(DisplayListPathEffect, UnknownEquals) {
+  const SkScalar TestDashes1[] = {4.0, 2.0};
+  sk_sp<SkPathEffect> sk_effect = SkDashPathEffect::Make(TestDashes1, 2, 0.0);
+  DlUnknownPathEffect effect1(sk_effect);
+  DlUnknownPathEffect effect2(sk_effect);
+  TestEquals(effect1, effect1);
+}
+
+TEST(DisplayListPathEffect, UnknownNotEquals) {
+  const SkScalar TestDashes1[] = {4.0, 2.0};
+  // Even though the effect is the same, it is a different instance
+  // and we cannot currently tell them apart because the Skia
+  // DashEffect::Make objects do not implement ==
+  DlUnknownPathEffect path_effect1(SkDashPathEffect::Make(TestDashes1, 2, 0.0));
+  DlUnknownPathEffect path_effect2(SkDashPathEffect::Make(TestDashes1, 2, 0.0));
+  TestNotEquals(path_effect1, path_effect2,
+                "SkDashPathEffect instance differs");
+}
+
+void testEquals(DlPathEffect* a, DlPathEffect* b) {
+  // a and b have the same nullness or values
+  ASSERT_TRUE(Equals(a, b));
+  ASSERT_FALSE(NotEquals(a, b));
+  ASSERT_TRUE(Equals(b, a));
+  ASSERT_FALSE(NotEquals(b, a));
+}
+
+void testNotEquals(DlPathEffect* a, DlPathEffect* b) {
+  // a and b do not have the same nullness or values
+  ASSERT_FALSE(Equals(a, b));
+  ASSERT_TRUE(NotEquals(a, b));
+  ASSERT_FALSE(Equals(b, a));
+  ASSERT_TRUE(NotEquals(b, a));
+}
+
+void testEquals(std::shared_ptr<const DlPathEffect> a, DlPathEffect* b) {
+  // a and b have the same nullness or values
+  ASSERT_TRUE(Equals(a, b));
+  ASSERT_FALSE(NotEquals(a, b));
+  ASSERT_TRUE(Equals(b, a));
+  ASSERT_FALSE(NotEquals(b, a));
+}
+
+void testNotEquals(std::shared_ptr<const DlPathEffect> a, DlPathEffect* b) {
+  // a and b do not have the same nullness or values
+  ASSERT_FALSE(Equals(a, b));
+  ASSERT_TRUE(NotEquals(a, b));
+  ASSERT_FALSE(Equals(b, a));
+  ASSERT_TRUE(NotEquals(b, a));
+}
+
+void testEquals(std::shared_ptr<const DlPathEffect> a,
+                std::shared_ptr<const DlPathEffect> b) {
+  // a and b have the same nullness or values
+  ASSERT_TRUE(Equals(a, b));
+  ASSERT_FALSE(NotEquals(a, b));
+  ASSERT_TRUE(Equals(b, a));
+  ASSERT_FALSE(NotEquals(b, a));
+}
+
+void testNotEquals(std::shared_ptr<const DlPathEffect> a,
+                   std::shared_ptr<const DlPathEffect> b) {
+  // a and b do not have the same nullness or values
+  ASSERT_FALSE(Equals(a, b));
+  ASSERT_TRUE(NotEquals(a, b));
+  ASSERT_FALSE(Equals(b, a));
+  ASSERT_TRUE(NotEquals(b, a));
+}
+
+TEST(DisplayListPathEffect, ComparableTemplates) {
```

Comment:
Oops, sorry. I think you copied this "ComparableTemplates" test from the mask filter unit tests. That test was the single test to confirm that the Equals and NotEquals helper methods worked and doesn't need to be copied to other unit tests.

You can delete this "ComparableTemplates" test and all of its helper functions as they don't really add anything.